returnn/torch/util/gradient_checkpoint.py

```diff
@@ -0,0 +1,577 @@
+
+    def maybe_store_amp_state(self, arg: Any):
+        """store AMP state if not yet stored for this device."""
+        if isinstance(arg, torch.Tensor):
+            device = arg.device
+        elif isinstance(arg, torch.device):
+            device = arg
+        else:
+            return
+        if device not in self.stored_device_amp_states:
+            self.stored_device_amp_states[device] = _get_dev_amp_state(device)
+
+    def maybe_map_raw_tensor_to_graph_tensor(self, tensor: torch.Tensor) -> Union[_GraphTensor, torch.Tensor]:
+        """raw tensor to graph tensor if available, otherwise return raw tensor."""
+        return self.graph_tensor_from_weak_raw_tensor.get(tensor, tensor)
+
+    def maybe_recompute(self):
+        """
+        Recompute.
+
+        Make sure that the recomputations happen in the correct order,
+        to get any random number generator state correct.
+
+        Note that we considered to have an API here which allowed to only recompute a subset of the ops.
+        It would still compute all from op idx 0 to some given op idx, but not the rest.
+        On subsequent calls, it would then continue from the last idx until again the requested op idx.
+        This works fine except of one important aspect: The RNG state.
+        If there are any other ops in between which use the RNG state, the RNG state would not be correct anymore.
+        So we cannot allow this. We must recompute all ops together right now.
+
+        However, we can at least remove the op from the list once it is computed.
+        So once any referenced tensor is not needed anymore, it can be garbage collected.
+        """
+        if not self.ops_to_be_recomputed:
+            return
+        with _reset_rng_states_scope(self.stored_device_rng_states), _reset_amp_states_scope(
+            self.stored_device_amp_states
+        ):
+            while self.ops_to_be_recomputed:
+                op = self.ops_to_be_recomputed.pop(0)
+                op.recompute()
+        self.stored_device_rng_states.clear()
+        self.stored_device_amp_states.clear()
+
+
+@dataclass
+class _GraphOp:
+    graph: _Graph
+    func: Any
+    args: Optional[Sequence[Union[_GraphTensor, Any]]]
+    kwargs: Optional[Dict[str, Union[_GraphTensor, Any]]]
+    out_flat_num: int
+    recomputed_out_flat: Optional[Sequence[torch.Tensor]] = None
+
+    def recompute(self):
+        """recompute, assuming all args are recomputed."""
+        args = pytree.tree_map_only(_GraphTensor, _GraphTensor.get_recomputed, self.args)
+        kwargs = pytree.tree_map_only(_GraphTensor, _GraphTensor.get_recomputed, self.kwargs)
+        out = self.func(*args, **kwargs)
+        out_flat, _ = pytree.tree_flatten(out)
+        assert len(out_flat) == self.out_flat_num
+        self.recomputed_out_flat = out_flat
+        # potentially free any referenced resources. we don't need them anymore.
+        self.args = None
+        self.kwargs = None
```

Comment:
`func` comes from `__torch_dispatch__`, so should only refer to any of the low-level aten native functions. I thought it might be helpful to keep the ref for debugging.